lib/pbench/server/api/resources/graphql_api.py

```diff
@@ -54,35 +57,199 @@ def post(self):
             self.logger.warning("Description not provided during metadata creation")
             abort(400, message="Please provide a description string")
         current_user_id = auth.current_user().id
+
         try:
-            # query GraphQL
-            query = f"""
-                    mutation {{ 
-                        createMetadata (input: {{config:"{config}", description:"{description}", user_id:{current_user_id}}}) {{ 
-                                metadata {{
-                                    id 
-                                    config 
-                                    description
-                                    }} 
-                                }} 
-                        }}
-                    """
-            result = schema.execute(query)
-        except Exception as e:
+            # Create a new metadata session
+            metadata_session = MetadataModel(
+                created=str(datetime.datetime.now()),
+                config=config,
+                description=description,
+                user_id=current_user_id
```

Comment:
If this is supposed to encompass what is proposed in https://github.com/distributed-system-analysis/pbench-dashboard/pull/117, then we need to be sure what entity sets the "created" and "updated" field times, the server or the browser.

In this new pbench-server way of doing things it should arguably be the server.  But we should reconcile this two PRs.

Reply:
Yes; we need to be careful and deliberate here about defining exactly what is "client metadata" and what is part of the server's metadata model.

What's being prototyped here is a good start, but it's just moving the server's GraphQL metadata model into the server, and I'm not sure that's really what we want going forward.

In particular, are "description", "config", and "updated" part of the server model, or simply part of an opaque client-owned metadata payload?

I can see an argument for the server keeping track of a timestamp for creation/changes. "description" and "config" are a bit specific.

I'd really prefer to be defining and building a more generic "client metadata" model that can support the dashboard's need to store the "saved session" data; but remember we also need to store the "favorite" data which is currently just on the browser, so we already have two "metadata keys" with different characteristics.

It seems to me that the basic extensible model is a simple key/value pair, with value representing a server-opaque client "blob" (whether a true uninterpreted binary blob or a JSON string). The contents of the value would be entirely up to the client; the server just stores it and retrieves it, updates it, or removes it on command.

So does the server timestamp create/update on the key/value pair, or is that part of the client blob the server doesn't need to understand? Same for "description". (I could see a value to a secondary API that defines the key values with a description string to help with a generic management dashboard view, for example. Do we need the server to manage a description for each key/value pair, though?)